Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.mm

```diff
@@ -0,0 +1,152 @@
+#include <CoreMotion/CoreMotion.h>
+
+#define ENABLE_STEP_COUNTER_LOGGING 1
+#if ENABLE_STEP_COUNTER_LOGGING
+#define STEP_COUNTER_LOG(...) NSLog(@"StepCounter - %@", [NSString stringWithFormat: __VA_ARGS__])
+#else
+#define STEP_COUNTER_LOG(...) {}
+#endif
+
+class iOSStepCounterWrapper
+{
+public:
+    typedef void (*OnDataReceived) (int deviceId, int numberOfSteps);
+
+    struct iOSStepCounterCallbacks
+    {
+        OnDataReceived dataReceived;
+    };
+
+
+    iOSStepCounterWrapper()
+    {
+        m_Pedometer = nullptr;
+        m_Device = -1;
+    }
+
+    void Enable(int deviceId, iOSStepCounterCallbacks* callbacks)
+    {
+        if (IsEnabled())
+        {
+            STEP_COUNTER_LOG(@"Was already enabled?");
+            if (m_Device != deviceId)
+                STEP_COUNTER_LOG(@"Enabling with different device id? Expected %d, was %d. Are you creating more than one iOSStepCounter", m_Device, deviceId);
+        }
+        m_Pedometer = [[CMPedometer alloc]init];
+        m_Callbacks = *callbacks;
+        m_Device = deviceId;
+        [m_Pedometer startPedometerUpdatesFromDate: [NSDate date] withHandler:^(CMPedometerData * _Nullable pedometerData, NSError * _Nullable error)
+        {
+            // Note: We need to call our callback on the same thread the Unity scripting is operating
```

Comment:
If we could get access to input module native api here, then no, it can be queued from any thread

Reply:
I'll try to explore, since dispatching like this - probably delays the input event by quite a lot